lib/resty/kafka/protocol/record.lua

```diff
@@ -0,0 +1,198 @@
+-- @author bzp2010 <bzp2010@apache.org>
+local function _message_set_v2_decode(resp, fetch_offset)
+    local messages = {}
+
+    -- RecordBatch decoder, refer to this documents
+    -- https://kafka.apache.org/documentation/#recordbatch
+    local base_offset = resp:int64() -- baseOffset
+    local batch_length = resp:int32() -- batchLength
+    local partition_leader_epoch = resp:int32() -- partitionLeaderEpoch
+    local magic_byte = resp:int8() -- magic
+    local crc = resp:int32() -- crc
+
+    -- Get all remaining message bodies by length for crc. The crc content
+    -- starts with attributes, so here we need to reduce its middle
+    -- three fields' length, include partition_leader_epoch (4 byte),
+    -- magic_byte (1 byte) and crc (4 byte).
+    local crc_content = resp:peek_bytes(resp.offset, batch_length - 4 - 1 - 4)
+    local calc_crc = crc32c(crc_content)
+    if crc ~= calc_crc then
+        return nil, nil, "crc checksum error"
+    end
+
+    -- TODO: support compressed Message Set
+    local attributes = resp:int16() -- attributes
+
+    local last_offset_delta = resp:int32() -- lastOffsetDelta
+    local last_offset = base_offset + last_offset_delta
+
+    -- If the last record's offset is also less than fetch's offset,
+    -- all outdated records are discarded.
+    if last_offset < fetch_offset then
+        resp:close()
+        return nil, nil, "all records outdated"
+    end
+
+    -- RecordBatch contains the timestamp starting value and the
+    -- maximum value of these records.
+    local first_timestamp = resp:int64() -- firstTimestamp
+    local max_timestamp = resp:int64() -- maxTimestamp
+
+    -- These fields are intended to support idempotent messages.
+    -- The features are NYI
+    local producer_id = resp:int64() -- producerId
+    local producer_epoch = resp:int16() -- producerEpoch
+    local base_sequence = resp:int32() -- baseSequence
+
+    local record_num = resp:int32() -- [records] array length
+
+    for i = 1, record_num do
+        messages[i] = {}
+
+        -- Record decoder, refer to this documents
+        -- https://kafka.apache.org/documentation/#record
+        local len = resp:varint()
+        local message_end = resp.offset + len
+
+        -- According to the protocol, only reserved.
+        local record_attributes = resp:int8()
+
+        -- Offset of this Record from RecordBatch's base value.
+        local timestamp_delta = resp:varlong()
+        local offset_delta = resp:varint()
+
+        -- The sixth bit of isControlBatch (bit 5) in attributes has a value of 1,
+        -- i.e. the current MessageSet (RecordBatch) is a ControlBatch.
+        -- !! Not process Conrtol Batch for now, they will be skipped !!
+        if band(attributes, 0x20) > 0 then
+            resp:varint() -- keyLength skipped
+            resp:int16() -- ControlBatch version skipped
+            resp:int16() -- ControlBatch type skipped
+            goto continue
```

Comment:
worth a warning/error log here? before ignoring it.

Reply:
changed, a info log added here